File: packages/agentboard/src/runtime/agents/tracker.ts
import type { AgentEvent } from "../contracts/agent";
import { TERMINAL_STATUSES } from "../contracts/agent";

const MAX_EVENT_TIMESTAMPS = 30;
const TERMINAL_PRUNE_MS = 5 * 60 * 1000;

const STATUS_PRIORITY: Record<string, number> = {
  running: 5,
  question: 4,
  error: 4,
  interrupted: 3,
  waiting: 2,
  done: 1,
  idle: 0,
};

export function instanceKey(agent: string, threadId?: string): string {
  return threadId ? `${agent}:${threadId}` : agent;
}

export class AgentTracker {
  // Outer key: session name, inner key: instance key (agent or agent:threadId)
  private instances = new Map<string, Map<string, AgentEvent>>();
  private eventTimestamps = new Map<string, number[]>();
  // Per-instance unseen tracking: "session\0instanceKey"
  private unseenInstances = new Set<string>();
  private active = new Set<string>();
  // Pinned instances: agents backed by a live pane process — exempt from pruning
  private pinnedKeys = new Map<string, Set<string>>(); // session → Set<instanceKey>

  private unseenKey(session: string, key: string): string {
    return `${session}\0${key}`;
  }

  /** Remove an instance and its unseen flag from a session's instance map. */
  private removeInstance(
    session: string,
    sessionInstances: Map<string, AgentEvent>,
    key: string,
  ): void {
    sessionInstances.delete(key);
    this.unseenInstances.delete(this.unseenKey(session, key));
  }

  applyEvent(event: AgentEvent, options?: { seed?: boolean }): void {
    const key = instanceKey(event.agent, event.threadId);

    // Store instance
    let sessionInstances = this.instances.get(event.session);
    if (!sessionInstances) {
      sessionInstances = new Map();
      this.instances.set(event.session, sessionInstances);
    }
    sessionInstances.set(key, event);

    // Track event timestamps
    let timestamps = this.eventTimestamps.get(event.session);
    if (!timestamps) {
      timestamps = [];
      this.eventTimestamps.set(event.session, timestamps);
    }
    timestamps.push(event.ts);
    if (timestamps.length > MAX_EVENT_TIMESTAMPS) {
      timestamps.splice(0, timestamps.length - MAX_EVENT_TIMESTAMPS);
    }

    // Per-instance unseen tracking
    // Seeded events always mark as unseen (they represent state from before the user connected)
    const ukey = this.unseenKey(event.session, key);
    if (TERMINAL_STATUSES.has(event.status)) {
      if (options?.seed || !this.active.has(event.session)) {
        this.unseenInstances.add(ukey);
      }
    } else {
      // Non-terminal status for this instance = user is interacting, mark seen
      this.unseenInstances.delete(ukey);
    }
  }

  /** Returns the most important agent state for backward compat */
  getState(session: string): AgentEvent | null {
    const sessionInstances = this.instances.get(session);
    if (!sessionInstances || sessionInstances.size === 0) return null;

    let best: AgentEvent | null = null;
    let bestPriority = -1;
    for (const event of sessionInstances.values()) {
      const p = STATUS_PRIORITY[event.status] ?? 0;
      if (p > bestPriority) {
        bestPriority = p;
        best = event;
      }
    }
    return best;
  }

  /** Returns all agent instances for a session, with unseen flag stamped */
  getAgents(session: string): AgentEvent[] {
    const sessionInstances = this.instances.get(session);
    if (!sessionInstances) return [];
    return [...sessionInstances.values()]
      .map((event) => {
        const key = instanceKey(event.agent, event.threadId);
        const isUnseen = this.unseenInstances.has(this.unseenKey(session, key));
        return isUnseen ? { ...event, unseen: true } : event;
      })
      .sort((a, b) => b.ts - a.ts);
  }

  /** Returns recent event timestamps for sparkline rendering */
  getEventTimestamps(session: string): number[] {
    return this.eventTimestamps.get(session) ?? [];
  }

  /** Clear unseen flags for every instance in a session, keeping the instances themselves
   * (pruneTerminal removes seen terminal instances after a timeout). */
  private clearUnseen(session: string): void {
    const sessionInstances = this.instances.get(session);
    if (!sessionInstances) return;
    for (const key of sessionInstances.keys()) {
      this.unseenInstances.delete(this.unseenKey(session, key));
    }
  }

  markSeen(session: string): boolean {
    if (!this.isUnseen(session)) return false;
    this.clearUnseen(session);
    return true;
  }

  dismiss(session: string, agent: string, threadId?: string): boolean {
    const sessionInstances = this.instances.get(session);
    if (!sessionInstances) return false;

    const key = instanceKey(agent, threadId);
    if (!sessionInstances.has(key)) return false;

    this.removeInstance(session, sessionInstances, key);
    if (sessionInstances.size === 0) {
      this.instances.delete(session);
    }
    return true;
  }

  pruneStuck(timeoutMs: number): void {
    const now = Date.now();
    for (const [session, sessionInstances] of this.instances) {
      for (const [key, event] of sessionInstances) {
        if (event.status === "running" && now - event.ts > timeoutMs) {
          if (this.isPinned(session, key)) continue;
          this.removeInstance(session, sessionInstances, key);
        }
      }
      if (sessionInstances.size === 0) {
        this.instances.delete(session);
      }
    }
  }

  /**
   * When multiple instances of the same agent share a paneId, keep only the most recent.
   * The currently-live instance is pinned by the pane scanner, so this only removes superseded
   * predecessors (e.g. a Claude Code session that was /exited and replaced in the same pane).
   */
  pruneSupersededByPane(): void {
    for (const [session, sessionInstances] of this.instances) {
      const groups = new Map<string, { key: string; ts: number }[]>();
      for (const [key, event] of sessionInstances) {
        if (!event.paneId) continue;
        const groupKey = `${event.paneId}\0${event.agent}`;
        const ts = event.details?.lastActivityAt ?? event.ts;
        const list = groups.get(groupKey) ?? [];
        list.push({ key, ts });
        groups.set(groupKey, list);
      }
      for (const list of groups.values()) {
        if (list.length < 2) continue;
        list.sort((a, b) => b.ts - a.ts);
        for (let i = 1; i < list.length; i++) {
          const k = list[i]!.key;
          if (this.isPinned(session, k)) continue;
          this.removeInstance(session, sessionInstances, k);
        }
      }
      if (sessionInstances.size === 0) {
        this.instances.delete(session);
      }
    }
  }

  /** Prune instances older than timeoutMs (by last activity) unless pinned, optionally
   * restricted to a single status. Shared by pruneStale (all statuses) and pruneIdle. */
  private pruneByAge(timeoutMs: number, onlyStatus?: AgentEvent["status"]): void {
    const now = Date.now();
    for (const [session, sessionInstances] of this.instances) {
      for (const [key, event] of sessionInstances) {
        if (onlyStatus && event.status !== onlyStatus) continue;
        if (this.isPinned(session, key)) continue;
        const lastSeen = event.details?.lastActivityAt ?? event.ts;
        if (now - lastSeen > timeoutMs) {
          this.removeInstance(session, sessionInstances, key);
        }
      }
      if (sessionInstances.size === 0) {
        this.instances.delete(session);
      }
    }
  }

  /** Auto-prune any instance whose last activity is older than timeoutMs, regardless of status. Skips pinned. */
  pruneStale(timeoutMs: number): void {
    this.pruneByAge(timeoutMs);
  }

  /**
   * Auto-prune "idle" instances older than timeoutMs unless pinned.
   * An agent only becomes idle when its process is gone or its journal stalled, so an
   * unpinned idle instance is a dead session (e.g. a Claude session left behind by /clear).
   * Live agents are pinned by the pane scanner and re-added via pane presence, so they survive.
   */
  pruneIdle(timeoutMs: number): void {
    this.pruneByAge(timeoutMs, "idle");
  }

  /** Auto-prune terminal instances older than timeout, but only if instance is not unseen or pinned */
  pruneTerminal(): void {
    const now = Date.now();
    for (const [session, sessionInstances] of this.instances) {
      for (const [key, event] of sessionInstances) {
        if (!TERMINAL_STATUSES.has(event.status)) continue;
        const ukey = this.unseenKey(session, key);
        if (this.unseenInstances.has(ukey)) continue; // Don't prune unseen — user hasn't looked yet
        if (this.isPinned(session, key)) continue; // Don't prune agents backed by live panes
        if (now - event.ts > TERMINAL_PRUNE_MS) {
          this.removeInstance(session, sessionInstances, key);
        }
      }
      if (sessionInstances.size === 0) {
        this.instances.delete(session);
      }
    }
  }

  isUnseen(session: string): boolean {
    // Session is unseen if any instance within it is unseen
    const sessionInstances = this.instances.get(session);
    if (!sessionInstances) return false;
    for (const key of sessionInstances.keys()) {
      if (this.unseenInstances.has(this.unseenKey(session, key))) return true;
    }
    return false;
  }

  getUnseen(): string[] {
    // Derive session-level unseen from per-instance tracking
    const sessions = new Set<string>();
    for (const ukey of this.unseenInstances) {
      sessions.add(ukey.split("\0")[0]!);
    }
    return [...sessions];
  }

  handleFocus(session: string): boolean {
    this.active.clear();
    this.active.add(session);

    const hadUnseen = this.isUnseen(session);
    if (hadUnseen) this.clearUnseen(session);
    return hadUnseen;
  }

  setActiveSessions(sessions: string[]): void {
    this.active.clear();
    for (const s of sessions) this.active.add(s);
  }

  /** Update the set of pinned instance keys for a session (live pane-backed agents). */
  setPinnedInstances(session: string | null, keys: string[]): void {
    this.pinnedKeys.clear();
    if (session && keys.length > 0) {
      this.pinnedKeys.set(session, new Set(keys));
    }
  }

  /** Update pinned instance keys for multiple sessions at once. */
  setPinnedInstancesMulti(keysBySession: Map<string, string[]>): void {
    this.pinnedKeys.clear();
    for (const [session, keys] of keysBySession) {
      if (keys.length > 0) {
        this.pinnedKeys.set(session, new Set(keys));
      }
    }
  }

  /** Check if an instance is pinned (backed by a live pane process). */
  isPinned(session: string, key: string): boolean {
    return this.pinnedKeys.get(session)?.has(key) ?? false;
  }
}

File: crates/tt-agentboard/src/tracker.rs
//! In-memory agent-instance state machine.
//!
//! Pure logic: every prune method takes an explicit `now_ms`, so tests are
//! deterministic. Insertion order is preserved with `IndexMap`/`IndexSet`.

use indexmap::{IndexMap, IndexSet};
use std::collections::{HashMap, HashSet};

use crate::types::AgentEvent;

const TERMINAL_PRUNE_MS: i64 = 5 * 60 * 1000;

/// The per-session instance-map key: `agent` or `agent:threadId`.
pub fn instance_key(agent: &str, thread_id: Option<&str>) -> String {
    match thread_id {
        Some(t) => format!("{agent}:{t}"),
        None => agent.to_string(),
    }
}

/// Tracks agent instances per session, their unseen state, pins, and prunes
/// dead/stale/terminal instances.
#[derive(Debug, Default)]
pub struct AgentTracker {
    /// session name → (instance key → latest event), insertion-ordered.
    instances: IndexMap<String, IndexMap<String, AgentEvent>>,
    /// Per-instance unseen tracking, keyed by `session\0instanceKey`.
    unseen_instances: IndexSet<String>,
    active: HashSet<String>,
    /// session → pinned instance keys (agents backed by a live pane process).
    pinned_keys: HashMap<String, HashSet<String>>,
}

fn unseen_key(session: &str, key: &str) -> String {
    format!("{session}\0{key}")
}

impl AgentTracker {
    pub fn new() -> Self {
        Self::default()
    }

    fn remove_instance(&mut self, session: &str, key: &str) {
        if let Some(inner) = self.instances.get_mut(session) {
            inner.shift_remove(key);
        }
        self.unseen_instances.shift_remove(&unseen_key(session, key));
    }

    /// Drop any session whose instance map is now empty.
    fn drop_if_empty(&mut self, session: &str) {
        if self.instances.get(session).is_some_and(IndexMap::is_empty) {
            self.instances.shift_remove(session);
        }
    }

    /// Record an event. `seed` marks pre-connection state, which always counts as
    /// unseen when terminal.
    pub fn apply_event(&mut self, event: AgentEvent, seed: bool) {
        let key = instance_key(&event.agent, event.thread_id.as_deref());
        let session = event.session.clone();
        let status = event.status;

        self.instances.entry(session.clone()).or_default().insert(key.clone(), event);

        let ukey = unseen_key(&session, &key);
        if status.is_terminal() {
            if seed || !self.active.contains(&session) {
                self.unseen_instances.insert(ukey);
            }
        } else {
            self.unseen_instances.shift_remove(&ukey);
        }
    }

    /// All instances for a session, unseen flag stamped, newest-first.
    pub fn get_agents(&self, session: &str) -> Vec<AgentEvent> {
        let Some(inner) = self.instances.get(session) else {
            return Vec::new();
        };
        let mut out: Vec<AgentEvent> = inner
            .values()
            .map(|event| {
                let key = instance_key(&event.agent, event.thread_id.as_deref());
                let mut ev = event.clone();
                if self.unseen_instances.contains(&unseen_key(session, &key)) {
                    ev.unseen = Some(true);
                }
                ev
            })
            .collect();
        // Stable sort by descending ts, so ties keep insertion order.
        out.sort_by_key(|e| std::cmp::Reverse(e.ts));
        out
    }

    fn clear_unseen(&mut self, session: &str) {
        let Some(inner) = self.instances.get(session) else {
            return;
        };
        let ukeys: Vec<String> = inner.keys().map(|k| unseen_key(session, k)).collect();
        for ukey in ukeys {
            self.unseen_instances.shift_remove(&ukey);
        }
    }

    /// Clear unseen flags for a session. Returns whether anything was unseen.
    pub fn mark_seen(&mut self, session: &str) -> bool {
        if !self.is_unseen(session) {
            return false;
        }
        self.clear_unseen(session);
        true
    }

    /// Remove a specific instance.
    pub fn dismiss(&mut self, session: &str, agent: &str, thread_id: Option<&str>) -> bool {
        let key = instance_key(agent, thread_id);
        let exists = self.instances.get(session).is_some_and(|inner| inner.contains_key(&key));
        if !exists {
            return false;
        }
        self.remove_instance(session, &key);
        self.drop_if_empty(session);
        true
    }

    /// Prune "running" instances older than `timeout_ms` (unless pinned).
    pub fn prune_stuck(&mut self, timeout_ms: i64, now_ms: i64) {
        let sessions: Vec<String> = self.instances.keys().cloned().collect();
        for session in sessions {
            let removable: Vec<String> = self.instances[&session]
                .iter()
                .filter(|(key, event)| {
                    event.status == crate::types::AgentStatus::Busy
                        && now_ms - event.ts > timeout_ms
                        && !self.is_pinned(&session, key)
                })
                .map(|(key, _)| key.clone())
                .collect();
            for key in removable {
                self.remove_instance(&session, &key);
            }
            self.drop_if_empty(&session);
        }
    }

    /// Prune instances whose last activity is older than `timeout_ms`, optionally
    /// restricted to one status; skips pinned.
    fn prune_by_age(
        &mut self,
        timeout_ms: i64,
        only_status: Option<crate::types::AgentStatus>,
        now_ms: i64,
    ) {
        let sessions: Vec<String> = self.instances.keys().cloned().collect();
        for session in sessions {
            let removable: Vec<String> = self.instances[&session]
                .iter()
                .filter(|(key, event)| {
                    if let Some(s) = only_status
                        && event.status != s
                    {
                        return false;
                    }
                    if self.is_pinned(&session, key) {
                        return false;
                    }
                    let last_seen =
                        event.details.as_ref().and_then(|d| d.last_activity_at).unwrap_or(event.ts);
                    now_ms - last_seen > timeout_ms
                })
                .map(|(key, _)| key.clone())
                .collect();
            for key in removable {
                self.remove_instance(&session, &key);
            }
            self.drop_if_empty(&session);
        }
    }

    /// Prune any instance whose last activity is older than `timeout_ms`.
    pub fn prune_stale(&mut self, timeout_ms: i64, now_ms: i64) {
        self.prune_by_age(timeout_ms, None, now_ms);
    }

    /// Prune "idle" instances older than `timeout_ms` unless pinned.
    pub fn prune_idle(&mut self, timeout_ms: i64, now_ms: i64) {
        self.prune_by_age(timeout_ms, Some(crate::types::AgentStatus::Idle), now_ms);
    }

    /// Prune terminal instances older than the terminal timeout, but only if seen
    /// and not pinned.
    pub fn prune_terminal(&mut self, now_ms: i64) {
        let sessions: Vec<String> = self.instances.keys().cloned().collect();
        for session in sessions {
            let removable: Vec<String> = self.instances[&session]
                .iter()
                .filter(|(key, event)| {
                    event.status.is_terminal()
                        && !self.unseen_instances.contains(&unseen_key(&session, key))
                        && !self.is_pinned(&session, key)
                        && now_ms - event.ts > TERMINAL_PRUNE_MS
                })
                .map(|(key, _)| key.clone())
                .collect();
            for key in removable {
                self.remove_instance(&session, &key);
            }
            self.drop_if_empty(&session);
        }
    }

    /// Whether any instance in the session is unseen.
    fn is_unseen(&self, session: &str) -> bool {
        let Some(inner) = self.instances.get(session) else {
            return false;
        };
        inner.keys().any(|key| self.unseen_instances.contains(&unseen_key(session, key)))
    }

    /// Set pinned instance keys for multiple sessions at once.
    pub fn set_pinned_instances_multi(&mut self, keys_by_session: &HashMap<String, Vec<String>>) {
        self.pinned_keys.clear();
        for (session, keys) in keys_by_session {
            if !keys.is_empty() {
                self.pinned_keys.insert(session.clone(), keys.iter().cloned().collect());
            }
        }
    }

    /// Whether an instance is pinned (backed by a live pane).
    pub fn is_pinned(&self, session: &str, key: &str) -> bool {
        self.pinned_keys.get(session).is_some_and(|s| s.contains(key))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{AgentEventDetails, AgentStatus};

    fn ev(session: &str, agent: &str, status: AgentStatus, ts: i64) -> AgentEvent {
        AgentEvent {
            agent: agent.into(),
            session: session.into(),
            status,
            ts,
            thread_id: None,
            thread_name: None,
            unseen: None,
            details: None,
        }
    }

    #[test]
    fn instance_key_with_and_without_thread() {
        assert_eq!(instance_key("claude", None), "claude");
        assert_eq!(instance_key("claude", Some("t1")), "claude:t1");
    }

    #[test]
    fn get_agents_sorted_newest_first_with_unseen_stamp() {
        let mut t = AgentTracker::new();
        // seed=true → terminal marked unseen.
        t.apply_event(ev("s", "old", AgentStatus::Complete, 10), true);
        t.apply_event(ev("s", "new", AgentStatus::Busy, 20), false);
        let agents = t.get_agents("s");
        assert_eq!(agents[0].agent, "new");
        assert_eq!(agents[1].agent, "old");
        assert_eq!(agents[1].unseen, Some(true));
        assert_eq!(agents[0].unseen, None);
    }

    #[test]
    fn non_terminal_event_clears_unseen() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s", "a", AgentStatus::Complete, 1), true);
        assert!(t.is_unseen("s"));
        // Same instance goes back to running → seen again.
        t.apply_event(ev("s", "a", AgentStatus::Busy, 2), false);
        assert!(!t.is_unseen("s"));
    }

    #[test]
    fn mark_seen_reports_only_the_first_clear() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s1", "a", AgentStatus::Complete, 1), true);
        t.apply_event(ev("s2", "b", AgentStatus::Error, 1), true);
        assert!(t.mark_seen("s1"));
        assert!(!t.mark_seen("s1")); // already seen
        assert!(t.mark_seen("s2"));
    }

    #[test]
    fn dismiss_removes_instance_and_empty_session() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s", "a", AgentStatus::Busy, 1), false);
        assert!(t.dismiss("s", "a", None));
        assert!(!t.dismiss("s", "a", None));
        assert!(t.get_agents("s").is_empty());
    }

    #[test]
    fn prune_stuck_removes_old_running_unless_pinned() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s", "a", AgentStatus::Busy, 0), false);
        t.apply_event(ev("s", "b", AgentStatus::Busy, 0), false);
        t.set_pinned_instances_multi(&HashMap::from([("s".to_string(), vec!["b".to_string()])]));
        t.prune_stuck(1000, 5000); // both are 5000ms old > 1000
        assert!(t.dismiss("s", "b", None)); // b survived (pinned)
        assert!(!t.dismiss("s", "a", None)); // a was pruned
    }

    #[test]
    fn prune_terminal_keeps_unseen_and_pinned() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s", "seen", AgentStatus::Complete, 0), true);
        t.mark_seen("s"); // clears both; the second lands unseen again below
        t.apply_event(ev("s", "unseen", AgentStatus::Complete, 0), true);
        t.prune_terminal(10 * 60 * 1000); // > TERMINAL_PRUNE_MS
        assert!(!t.dismiss("s", "seen", None)); // seen terminal pruned
        assert!(t.dismiss("s", "unseen", None)); // unseen kept
    }

    #[test]
    fn prune_idle_only_targets_idle() {
        let mut t = AgentTracker::new();
        t.apply_event(ev("s", "idle", AgentStatus::Idle, 0), false);
        t.apply_event(ev("s", "run", AgentStatus::Busy, 0), false);
        t.prune_idle(1000, 5000);
        assert!(!t.dismiss("s", "idle", None)); // idle pruned
        assert!(t.dismiss("s", "run", None)); // running kept
    }

    #[test]
    fn prune_by_age_uses_last_activity_when_present() {
        let mut t = AgentTracker::new();
        let mut e = ev("s", "a", AgentStatus::Busy, 0);
        e.details = Some(AgentEventDetails { last_activity_at: Some(4500), ..Default::default() });
        t.apply_event(e, false);
        // event ts is 0 (very old) but lastActivityAt is recent → not stale.
        t.prune_stale(1000, 5000);
        assert!(t.dismiss("s", "a", None));
    }
}
